src/sentry/services/hybrid_cloud/organization_mapping/impl.py

```diff
@@ -0,0 +1,74 @@
+from dataclasses import fields
+from typing import Any, List, Optional
+
+from django.db import IntegrityError, transaction
+
+from sentry.models.organizationmapping import OrganizationMapping
+from sentry.models.user import User
+from sentry.services.hybrid_cloud.organization_mapping import (
+    APIOrganizationMapping,
+    OrganizationMappingService,
+)
+
+
+class DatabaseBackedOrganizationMappingService(OrganizationMappingService):
+    def create(
+        self,
+        *,
+        user: User,
+        organization_id: int,
+        slug: str,
+        region_name: str,
+        idempotency_key: Optional[str] = "",
+        # There's only a customer_id when updating an org slug
+        customer_id: Optional[str] = None,
+    ) -> APIOrganizationMapping:
+        try:
+            with transaction.atomic():
+                # Creating an identical mapping should succeed, even if a record already exists
+                # with this slug. We allow this IFF the idempotency key is identical.
+                org_mapping = OrganizationMapping.objects.create(
+                    organization_id=organization_id,
+                    slug=slug,
+                    customer_id=customer_id,
+                    idempotency_key=idempotency_key,
+                    region_name=region_name,
+                )
+                return self.serialize_organization_mapping(org_mapping)
+        except IntegrityError:
+            pass
+
+        # If we got here, the slug already exists
+        if idempotency_key != "":
+            try:
+                existing_mapping = OrganizationMapping.objects.get(
+                    slug=slug, idempotency_key=idempotency_key
+                )
+                existing_mapping.organization_id = organization_id
+                existing_mapping.customer_id = customer_id
+                existing_mapping.region_name = region_name
+                existing_mapping.save()
```

Comment:
We might want a transaction and to use `select_for_update()` here to ensure we don't get hit with concurrent updates.